Stop reporting exceptions in custom transformations as internal errors

jackkoenig opened the pull request:
Instead, just forward the exception

chick commented on src/main/scala/firrtl/Compiler.scala:
Do you need lines 480-481?
Seems like the only thing you are actually catching is the `isCustomTransform`, all other exceptions will bubble up either way.

jackkoenig replied:
I was worried about what would happen if we wrapped a ControlThrowable in a CustomTransformException, but since the Driver just forwards it, it'll do the right thing. I'll fix that.

I'm also going to make `isCustomTransform` a `def` since it only needs to be calculated once and in the rare case of an exception.

jackkoenig commented:
One potential problem with this is that it treats optional built-in transforms as Custom Transforms, so things like ReplSeqMems will be treated in the same way.

We implement this differently in a couple of ways:
1. Check if the transform is defined in the `firrtl` package
1. Create a whitelist of transforms (brittle and annoying to maintain)

I'm leaning toward 1. now. I had thought about that originally but I thought this approach would be more robust. While it is more robust, in my opinion it still has too many false negatives.

seldridge commented:
Is there anything that precludes a user building a library and putting it in the FIRRTL package? Does that matter?

chick commented:
AFAIK there isn't but what we should probably be doing is naming our packages properly as 
<umbrella>.<organizations>.firrtl
Probably too late for 3.2.x

jackkoenig commented:
I don't think there is anyway to prevent that. Of course a whitelist could prevent errors in such transforms from being reported as internal errors. I think it doesn't really matter though, if you inject your code into other people's packages you're always asking for trouble.

jackkoenig commented:
Per chisel things, I actually don't want this to treat things in `com.mycoolcompany.firrtl` as internal. Rather, only things defined in `firrtl.whatever`. I do grant that it's a problem for transforms defined in Chisel; however, there currently isn't a single one.

jackkoenig commented:
@chick @seldridge thoughts on the updated implementation?  Unfortunately I don't think there's a more robust to get the "top package" in which a class is defined other than splitting the package String by `.` and then loading the package named by the head of the resulting array. It looks like `package java` and `package java.lang` are two completely unrelated packages at runtime so I did some stringly-typed magic to determine the top-level package.

jackkoenig commented:
Okay this fails with a NullPointerException if the transform is defined in the root package, need to fix that

jackkoenig commented:
I guess I think it's likely that more and more build-in transforms will be run with that same mechanism and kind of the whole idea of the "Internal Error" Exception is to let the user know that this error is unhandled and is our fault. Relatedly, we should probably stop reporting user error as exceptions with useless stack traces and instead report the errors and exit

azidar commented:
So, should we merge this?

jackkoenig commented:
The implementation *could* be cleaner, but assuming it still works, I think this is very desirable. Thoughts @seldridge? The main alternative (a twist on Schuyler's proposal above), is to make all built-in transforms extend some special trait (eg. `BuiltInTransform`) so that we know what is internal and what is custom. Instead, this PR is using the package name, looking for "firrtl" which is a bit less clean but 🤷‍♂️

jackkoenig commented:
Fixed test and rebased (because while it's not a huge deal, the Mergify squash and merge results in intermediate commit messages being kept and I didn't want them from here). This is still fine to rebase and merge, it just won't have intermediate messages

seldridge commented:
FWIW, one of the gates on mergify is that the commit has to have the `Please Merge` tag. Without that, you're in complete control including to do a GitHub-based squash and merge.

jackkoenig commented:
Ah true, well in any case I cleaned this one up for Mergify but might just do that in the future :-P
